### DISCUSSION TITLE
After the Promise.all method is executed, the generated connection has not been released

### DISCUSSION THREAD

**Discussion by liyofx** [prisma/prisma#23809] [Q&A]
### Bug description

When using the Promise.all method for concurrent writing, the connection generated is not released after the relevant code execution is completed

### How to reproduce

```shell
➜  prisma-app bun index.ts
before connections: 3n
Done...
[25.58ms] exec
current_connections: 12n , users: 10
current_connections: 12n , users: 10
current_connections: 12n , users: 10
current_connections: 12n , users: 10
current_connections: 12n , users: 10
current_connections: 12n , users: 10
```

### Prisma information

```prisma
// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")  //postgresql://z:@localhost:5432/test
}

model User {
  id    Int     @id
  email String  @unique
  name  String?
}
```

```ts
import { PrismaClient } from "@prisma/client";

const prisma = new PrismaClient();

await (async () => {
  await prisma.user.deleteMany()
  var connections = await prisma.$queryRaw`SELECT count(*) AS current_connections FROM pg_stat_activity WHERE datname='test';`.then((value: any) => value[0].current_connections)
  console.log('before connections:', connections);
  setInterval(async () => {
    let connections = await prisma.$queryRaw`SELECT count(*) AS current_connections FROM pg_stat_activity WHERE datname='test';`.then((value: any) => value[0].current_connections)
    console.log('current_connections:', connections, ', users:', await prisma.user.count());
  }, 1000)
  console.time('exec');
  await insertData()
  console.timeEnd('exec');
})();

async function insertData() {
  var arr = Array.from({ length: 10 })
  await Promise.all(arr.map((v, index) => {
    var data = {
      id: index,
      name: "John Dough",
      email: `john-${Math.random()}-${Math.random()}-${index}@example.com`,
    }
    // return prisma.user.create({ data })
    return prisma.user.upsert({ create: data, update: data, where: { id: index } })
  }))
  console.log('Done...');
}
```

### Environment & setup

- OS: macOS 13.4.1
- Database: PostgreSQL 16.1 (Homebrew) on aarch64-apple-darwin22.6.0, compiled by Apple clang version 14.0.3 (clang-1403.0.22.14.1), 64-bit
- Node.js version: v21.6.1

### Prisma Version

```
Environment variables loaded from .env
prisma                  : 5.12.1
@prisma/client          : 5.12.1
Computed binaryTarget   : darwin-arm64
Operating System        : darwin
Architecture            : arm64
Node.js                 : v21.6.1
Query Engine (Node-API) : libquery-engine 473ed3124229e22d881cb7addf559799debae1ab (at node_modules/@prisma/engines/libquery_engine-darwin-arm64.dylib.node)
Schema Engine           : schema-engine-cli 473ed3124229e22d881cb7addf559799debae1ab (at node_modules/@prisma/engines/schema-engine-darwin-arm64)
Schema Wasm             : @prisma/prisma-schema-wasm 5.12.0-21.473ed3124229e22d881cb7addf559799debae1ab
Default Engines Hash    : 473ed3124229e22d881cb7addf559799debae1ab
Studio                  : 0.499.0
```

**Comment by SevInf** [comment]
Prisma does not close the connections after using them but puts them back into the pool to use later. You can control the size and other connection pool parameters by following [our docs](https://www.prisma.io/docs/orm/prisma-client/setup-and-configuration/databases-connections/connection-pool).

If you are getting "Too many connections" error from the database, it most likely means that the connection limit in your database is lower than Prisma's pool size. You can adjust the pool size by using `connection_limit` query string parameter in the connection string.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
